lib/ethash/ethash.cpp

```diff
@@ -428,29 +429,35 @@ ethash_hash256 ethash_calculate_epoch_seed(int epoch_number) noexcept
 
 int ethash_calculate_light_cache_num_items(int epoch_number) noexcept
 {
-    static constexpr int item_size = sizeof(hash512);
-    static constexpr int num_items_init = light_cache_init_size / item_size;
-    static constexpr int num_items_growth = light_cache_growth / item_size;
+    constexpr int item_size = sizeof(hash512);
+    constexpr int num_items_init = light_cache_init_size / item_size;
+    constexpr int num_items_growth = light_cache_growth / item_size;
     static_assert(
         light_cache_init_size % item_size == 0, "light_cache_init_size not multiple of item size");
     static_assert(
         light_cache_growth % item_size == 0, "light_cache_growth not multiple of item size");
 
+    if (epoch_number > max_epoch_number)
+        return 0;
+
```

Comment:
I think this might cause a cascading problem in create_epoch_context around here
```
   const size_t alloc_size = context_alloc_size + light_cache_size + full_dataset_size;
   char* const alloc_data = static_cast<char*>(std::calloc(1, alloc_size));
    if (!alloc_data)
        return nullptr;  // Signal out-of-memory by returning null pointer.

    hash512* const light_cache = reinterpret_cast<hash512*>(alloc_data + context_alloc_size);
```
If beyond max epoch then `alloc_size == 512+0+0` which means
`hash512* const light_cache = reinterpret_cast<hash512*>(alloc_data + context_alloc_size);` reinterpret a location beyond previous allocation

Reply:
The `create_epoch_context()` has the same check and returns `nullptr`. Now I also added a check for `epoch_number < 0`.

The `calculate_X_num_items()` have the check duplicated because they are public functions.

Reply:
Right sorry. Was comparing master branch while reading PR branch. My bad